hive/runners/multi_agent_loop.py

```diff
@@ -1,61 +1,111 @@
-        logger,
-        experiment_manager,
-        train_steps,
-        test_frequency,
-        test_episodes,
-        stack_size,
-        self_play,
-        max_steps_per_episode=27000,
+        environment: BaseEnv,
+        agents: List[Agent],
+        loggers: List[ScheduledLogger],
+        experiment_manager: Experiment,
+        train_steps: int,
+        num_agents: int,
+        eval_environment: BaseEnv = None,
+        test_frequency: int = -1,
+        test_episodes: int = 1,
+        stack_size: int = 1,
+        self_play: bool = False,
+        max_steps_per_episode: int = 1e9,
+        seed: int = None,
     ):
-        """Initializes the Runner object.
+        """Initializes the MultiAgentRunner object.
 
         Args:
             environment (BaseEnv): Environment used in the training loop.
-            agents (list[Agent]): List of agents that interact with the environment
-            logger (ScheduledLogger): Logger object used to log metrics.
+            agent (Agent): Agent that will interact with the environment
+            loggers (List[ScheduledLogger]): List of loggers used to log metrics.
             experiment_manager (Experiment): Experiment object that saves the state of
                 the training.
             train_steps (int): How many steps to train for. If this is -1, there is no
                 limit for the number of training steps.
+            num_agents (int): Number of agents running in this multiagent experiment.
+            eval_environment (BaseEnv): Environment used to evaluate the agent. If
+                None, the ``environment`` parameter (which is a function) is
+                used to create a second environment.
             test_frequency (int): After how many training steps to run testing
                 episodes. If this is -1, testing is not run.
-            test_episodes (int): How many episodes to run testing for.
+            test_episodes (int): How many episodes to run testing for duing each test
+                phase.
             stack_size (int): The number of frames in an observation sent to an agent.
+            self_play (bool): Whether this multiagent experiment is run in
+                self-play mode. In this mode, only the first agent in the list
+                of agents provided in the config is created. This agent performs
+                actions for each player in the multiagent environment.
             max_steps_per_episode (int): The maximum number of steps to run an episode
                 for.
+            seed (int): Seed used to set the global seed for libraries used by
+                Hive and seed the :py:class:`~hive.utils.utils.Seeder`.
         """
+        if seed is not None:
+            utils.seeder.set_global_seed(seed)
+        if eval_environment is None:
+            eval_environment = environment
+        environment = environment()
+        eval_environment = eval_environment() if test_frequency != -1 else None
```

Comment:
Why is this line repeated both here and in the `single_agent_loop`? 

The logic behind the `eval_environment` is the same for all runners, is it not? It's either provided in the `__init__` or it should be the same as *train_env*.

Reply:
Logic is two-fold:
1. Allow more flexibility for people who want to create their custom runners that have custom behaviors relating to evaluation environments
2. Make the behavior of how we create these environments more visible for anyone using SingleAgentRunner